### PR TITLE
[InstCombine] Use zext's nneg flag for icmp folding

### LINKED ISSUE #55013: Instruction pmaxsw not generated
The code generated by 'x86-64 clang 13.0.1' in /O3 will generate pmaxsw instructions. With 'x86-64 clang 14.0.0' and 'x86-64 clang (trunk)', the pmaxw is not generated, and the code size is slightly bigger.

```c
short vecreduce_smax_v2i16(int n, short* v)
{
  short p = 0;
  for (int i = 0; i < n; ++i)
    p = p < v[i] ? v[i] : p;
  return p;
}
```

### PR DESCRIPTION
**Pull request by leo-ard**
This PR fixes https://github.com/llvm/llvm-project/issues/55013 : the max intrinsics is not generated for this simple loop case : https://godbolt.org/z/hxz1xhMPh. This is caused by a ICMP not being folded into a select, thus not generating the max intrinsics.

For the story :

Since LLVM 14, SCCP pass got smarter by folding sext into zext for positive ranges : https://reviews.llvm.org/D81756. After this change, InstCombine was sometimes unable to fold ICMP correctly as both of the arguments pointed to mismatched zext/sext. To fix this, @rotateright implemented this fix : https://reviews.llvm.org/D124419 that tries to resolve the mismatch by knowing if the argument of a zext is positive (in which case, it is like a sext) by using ValueTracking, however ValueTracking is not smart enough to infer that the value is positive in some cases. Recently, @nikic implemented #67982 which keeps the information that a zext is non-negative. This PR simply uses this information to do the folding accordingly.

TLDR : This PR uses the recent nneg tag on zext to fold the icmp accordingly in instcombine.

This PR also contains test cases for sext/zext folding with InstCombine as well as a x86 regression tests for the max/min case.

### CODE REVIEW

File: clang/test/CodeGen/X86/min_max.c [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// RUN: %clang_cc1 %s -O2 -triple=x86_64-apple-darwin -emit-llvm -o - | FileCheck %s
```
Comment (nikic) [inline clang/test/CodeGen/X86/min_max.c:1]: We don't test end-to-end codegen with clang. If you want to test this, you should take the unoptimized clang IR and create a test in the llvm/test/PhaseOrdering directory using it.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 2]
```diff
@@ -5587,11 +5587,17 @@ Instruction *InstCombinerImpl::foldICmpWithZextOrSext(ICmpInst &ICmp) {
         return new ICmpInst(ICmp.getPredicate(), Builder.CreateOr(X, Y),
                             Constant::getNullValue(X->getType()));
 
+      // Treat "zext nneg" as "sext"
+      bool IsNonNeg0 = isa<PossiblyNonNegInst>(ICmp.getOperand(0));
+      bool IsNonNeg1 = isa<PossiblyNonNegInst>(ICmp.getOperand(1));
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5592]: This only checks whether the instruction *can* have an nneg flag, not whether it has it. You need to check `isNonNeg()` for that.

File: llvm/test/Transforms/InstCombine/icmp-fold-with-cast.ll [hunk 3]
```diff
@@ -0,0 +1,185 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --tool ./bin/opt --version 3
+; See PRXXX for more details
+; RUN-./bin/opt: opt < %s -S -passes=ipsccp | FileCheck %s
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/icmp-fold-with-cast.ll:3]: This is an ipsccp test in the InstCombine directory?

File: clang/test/CodeGen/X86/min_max.c (hunk 1 above)
Comment (leo-ard) [reply clang/test/CodeGen/X86/min_max.c]: Out of curiosity, why is end-to-end codegen with clang not used ? Is it too unstable ?

File: llvm/test/Transforms/InstCombine/icmp-fold-with-cast.ll (hunk 3 above)
Comment (leo-ard) [reply llvm/test/Transforms/InstCombine/icmp-fold-with-cast.ll]: yep, a typo on my end

**Comment by leo-ard** [comment]
Thanks for taking the time to review the PR. I just added another test in PhaseOrdering to make sure that the min/max intrinsics are generated

**Comment by leo-ard** [comment]
@nikic ping:)

File: llvm/test/Transforms/PhaseOrdering/min_max_loop.ll [hunk 4]
```diff
@@ -0,0 +1,126 @@
+; RUN: opt < %s --O3 -S | FileCheck %s
+
+define signext i16 @vecreduce_smax_v2i16(i32 noundef %0, ptr noundef %1) #0 {
```
Comment (dtcxzyw) [inline llvm/test/Transforms/PhaseOrdering/min_max_loop.ll:3]: Could you please reduce the test case and move it into `InstCombine`?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 5]
```diff
@@ -5587,11 +5587,20 @@ Instruction *InstCombinerImpl::foldICmpWithZextOrSext(ICmpInst &ICmp) {
...
+      bool IsNonNeg0 = NonNegInst0 && NonNegInst0->hasNonNeg();
+      bool IsNonNeg1 = NonNegInst1 && NonNegInst1->hasNonNeg();
+
       // If we have mismatched casts, treat the zext of a non-negative source as
       // a sext to simulate matching casts. Otherwise, we are done.
       // TODO: Can we handle some predicates (equality) without non-negative?
-      if ((IsZext0 && isKnownNonNegative(X, DL, 0, &AC, &ICmp, &DT)) ||
-          (IsZext1 && isKnownNonNegative(Y, DL, 0, &AC, &ICmp, &DT)))
+      if ((IsZext0 &&
+           (IsNonNeg0 || isKnownNonNegative(X, DL, 0, &AC, &ICmp, &DT))) ||
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5601]: Why not check the `nneg` flag in `isKnownNonNegative`?

File: llvm/test/Transforms/InstCombine/icmp-fold-with-cast.ll [hunk 6]
```diff
@@ -0,0 +1,175 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 3
+; See PR-70845 for more details
+; RUN: opt < %s -S -passes=instcombine | FileCheck %s
+
+
+define signext i32 @sext_sext(i16 %x, i16 %y) {
```
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/icmp-fold-with-cast.ll:6]: Please drop unused `signext` and `noundef` flags.

File: llvm/test/Transforms/InstCombine/icmp-fold-with-cast.ll [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+cond.end:                                         ; preds = %cond.false, %cond.true
+  %cond = phi i32 [ %conv1, %cond.true ], [ %conv, %cond.false ]
+  %conv8 = trunc i32 %cond to i16
+  %inc = add nsw i32 %i.0, 1
+  br label %for.cond
+}
+
+
+
+define signext i16 @sext_and_zext_positive(i16 %x) {
```
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/icmp-fold-with-cast.ll:133]: It looks like an infinite loop without side effects. It is a UB.

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp (hunk 5 above)
Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5601]: After rebasing over https://github.com/llvm/llvm-project/commit/5918f62301788b53e7d3a23f3203c483e9d4d791 it's possible to drop the isKnownNonNegative() call here entirely. We should only check for zext nneg.

File: llvm/test/Transforms/PhaseOrdering/min_max_loop.ll (hunk 4 above)
Comment (nikic) [reply llvm/test/Transforms/PhaseOrdering/min_max_loop.ll:3]: I think the point of these tests is to check interaction between IPSCCP and InstCombine, so a PhaseOrdering test makes sense.

However, in the current form these tests are too hard to understand. I would suggest getting the IR again with `-fno-discard-value-names` and then running `opt -S -passes=sroa` over it to convert it into SSA form.

File: llvm/test/Transforms/InstCombine/icmp-fold-with-cast.ll (hunk 6 above)
Comment (nikic) [reply llvm/test/Transforms/InstCombine/icmp-fold-with-cast.ll:6]: Generally the tests in this file are not right for InstCombine. We should just directly check the min/max pattern with zext nneg there, nothing more. Preferably directly next to the existing tests for the transform (you can find them by commenting out the transform and seeing which tests break).

File: llvm/test/Transforms/InstCombine/icmp-fold-with-cast.ll (hunk 7 above)
Comment (leo-ard) [reply llvm/test/Transforms/InstCombine/icmp-fold-with-cast.ll]: I think you are right. I removed the test

File: llvm/test/Transforms/InstCombine/icmp-ext-ext.ll [hunk 8]
```diff
@@ -247,6 +355,19 @@ define i1 @sext_zext_uge_op0_wide(i16 %x, i8 %y) {
...
+; CHECK-NEXT:    ret i1 [[C]]
+;
+  %a = sext i16 %x to i32
+  %b = zext nneg i8 %y to i32
+  %c = icmp uge i32 %a, %b
+  ret i1 %c
+}
+
 define i1 @zext_sext_sgt_known_nonneg(i8 %x, i8 %y) {
 ; CHECK-LABEL: @zext_sext_sgt_known_nonneg(
```
Comment (leo-ard) [inline llvm/test/Transforms/InstCombine/icmp-ext-ext.ll:372]: Every test in this test-file has an `known_nonneg` variant for mismatched zext/sext. I was wondering on the relevance of this variant as we now have the nneg flag on zext. It would make more sense to have specialized tests for zext -> zext nneg folding with value-tracking and another test for icmp with zext nneg/sext folding into icmp (what I just added). FYI, this test dates from this PR, when there where no nneg flag : https://reviews.llvm.org/D124419.

File: llvm/test/Transforms/InstCombine/icmp-fold-with-cast.ll (hunk 6 above)
Comment (leo-ard) [reply llvm/test/Transforms/InstCombine/icmp-fold-with-cast.ll]: The technique of commenting and seeing which tests crashed worked really well. Thanks for the trick

File: llvm/test/Transforms/PhaseOrdering/min_max_loop.ll [hunk 9]
```diff
@@ -0,0 +1,145 @@
+; RUN: opt < %s --O3 -S | FileCheck %s
```
Comment (dtcxzyw) [inline llvm/test/Transforms/PhaseOrdering/min_max_loop.ll:1]: Could you please:
1. Convert this IR into SSA form as @nikic mentioned.
2. Drop unused attributes.
3. Re-generate tests with `llvm/utils/update_test_checks.py`.

Comment (dtcxzyw) [inline llvm/test/Transforms/PhaseOrdering/min_max_loop.ll:1]:
```suggestion
; RUN: opt < %s -O3 -S | FileCheck %s
```

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/test/Transforms/PhaseOrdering/min_max_loop.ll (hunk 9 above)
Comment (dtcxzyw) [reply llvm/test/Transforms/PhaseOrdering/min_max_loop.ll:1]:
```
define i16 @vecreduce_smin_v2i16(i32 %n, ptr %v) {
entry:
  br label %for.cond

for.cond:
  %p.0 = phi i16 [ 0, %entry ], [ %conv8, %for.inc ]
  %i.0 = phi i32 [ 0, %entry ], [ %inc, %for.inc ]
  %cmp = icmp slt i32 %i.0, %n
  br i1 %cmp, label %for.body, label %for.end

for.body:
  %conv = sext i16 %p.0 to i32
  %idxprom = sext i32 %i.0 to i64
  %arrayidx = getelementptr inbounds i16, ptr %v, i64 %idxprom
  %0 = load i16, ptr %arrayidx, align 2
  %conv1 = sext i16 %0 to i32
  %cmp2 = icmp sgt i32 %conv, %conv1
  br i1 %cmp2, label %cond.true, label %cond.false

cond.true:
  %idxprom4 = sext i32 %i.0 to i64
  %arrayidx5 = getelementptr inbounds i16, ptr %v, i64 %idxprom4
  %1 = load i16, ptr %arrayidx5, align 2
  %conv6 = sext i16 %1 to i32
  br label %cond.end

cond.false:
  %conv7 = sext i16 %p.0 to i32
  br label %cond.end

cond.end:
  %cond = phi i32 [ %conv6, %cond.true ], [ %conv7, %cond.false ]
  %conv8 = trunc i32 %cond to i16
  br label %for.inc

for.inc:
  %inc = add nsw i32 %i.0, 1
  br label %for.cond

for.end:
  ret i16 %p.0
}

define i16 @vecreduce_smax_v2i16(i32 %n, ptr %v) {
entry:
  br label %for.cond

for.cond:
  %p.0 = phi i16 [ 0, %entry ], [ %conv8, %for.inc ]
  %i.0 = phi i32 [ 0, %entry ], [ %inc, %for.inc ]
  %cmp = icmp slt i32 %i.0, %n
  br i1 %cmp, label %for.body, label %for.end

for.body:
  %conv = sext i16 %p.0 to i32
  %idxprom = sext i32 %i.0 to i64
  %arrayidx = getelementptr inbounds i16, ptr %v, i64 %idxprom
  %0 = load i16, ptr %arrayidx, align 2
  %conv1 = sext i16 %0 to i32
  %cmp2 = icmp slt i32 %conv, %conv1
  br i1 %cmp2, label %cond.true, label %cond.false

cond.true:
  %idxprom4 = sext i32 %i.0 to i64
  %arrayidx5 = getelementptr inbounds i16, ptr %v, i64 %idxprom4
  %1 = load i16, ptr %arrayidx5, align 2
  %conv6 = sext i16 %1 to i32
  br label %cond.end

cond.false:
  %conv7 = sext i16 %p.0 to i32
  br label %cond.end

cond.end:
  %cond = phi i32 [ %conv6, %cond.true ], [ %conv7, %cond.false ]
  %conv8 = trunc i32 %cond to i16
  br label %for.inc

for.inc:
  %inc = add nsw i32 %i.0, 1
  br label %for.cond

for.end:
  ret i16 %p.0
}

```

File: llvm/test/Transforms/InstCombine/icmp-ext-ext.ll (hunk 8 above)
Comment (nikic) [reply llvm/test/Transforms/InstCombine/icmp-ext-ext.ll:372]: Yeah, these tests are largely redundant now. We could delete them, but I think it's also okay to keep them to show the combination of nneg inference and this fold.

Comment (leo-ard) [reply llvm/test/Transforms/InstCombine/icmp-ext-ext.ll:372]: Ok I'll leave them then

File: llvm/test/Transforms/PhaseOrdering/min_max_loop.ll (hunk 9 above)
Comment (leo-ard) [reply llvm/test/Transforms/PhaseOrdering/min_max_loop.ll:1]: How did you get the SSA form ? I wasn't able to do it on my side. Here are the commands that I did : 
```bash
> build_release/bin/clang -S -emit-llvm min_max.c -fno-discard-value-names -o min_max.ll
> build_release/bin/opt -S -passes=sroa min_max.ll > min_max2.ll
> head -n 20 min_max2.ll
; ModuleID = 'min_max.ll'
source_filename = "min_max.c"
target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128"
target triple = "arm64-apple-macosx14.0.0"

; Function Attrs: noinline nounwind optnone ssp uwtable(sync)
define signext i16 @vecreduce_smin_v2i16(i32 noundef %n, ptr noundef %v) #0 {
entry:
  %n.addr = alloca i32, align 4
  %v.addr = alloca ptr, align 8
  %p = alloca i16, align 2
  %i = alloca i32, align 4
  store i32 %n, ptr %n.addr, align 4
  store ptr %v, ptr %v.addr, align 8
  store i16 0, ptr %p, align 2
  store i32 0, ptr %i, align 4
  br label %for.cond

for.cond:                                         ; preds = %for.inc, %entry
  %0 = load i32, ptr %i, align 4
```

Comment (nikic) [reply llvm/test/Transforms/PhaseOrdering/min_max_loop.ll:1]: You need to use `-O2 -Xclang -disable-llvm-optzns`, or manually drop the `optnone` attributes.

File: llvm/test/Transforms/PhaseOrdering/min_max_loop.ll [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+;; {
+;;   short p = 0;
+;;   for (int i = 0; i < n; ++i)
+;;     p = p < v[i] ? v[i] : p;
+;;   return p;
+;; }
+
+define i16 @vecreduce_smin_v2i16(i32 %n, ptr %v) {
+; CHECK-LABEL: define i16 @vecreduce_smin_v2i16(
+; CHECK:    @llvm.smin.v2i16
```
Comment (leo-ard) [inline llvm/test/Transforms/PhaseOrdering/min_max_loop.ll:24]: I didn't keep all the checks generated by `llvm/utils/update_test_checks.py` as in this tests, we only want to know if the intrinsics @llvm.smin.v2i16 has been generated or not. Let me know if you think I should leave all the checks

Comment (dtcxzyw) [reply llvm/test/Transforms/PhaseOrdering/min_max_loop.ll:23]: Oh, it makes sense. Just checking `smin/smax` intrinsics is OK.

File: llvm/test/Transforms/PhaseOrdering/min_max_loop.ll [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+for.inc:                                          ; preds = %cond.end
+  %inc = add nsw i32 %i.0, 1
+  br label %for.cond
+
+for.end:                                          ; preds = %for.cond
+  ret i16 %p.0
+}
+
+define signext i16 @vecreduce_smax_v2i16(i32 noundef %n, ptr noundef %v) {
```
Comment (dtcxzyw) [inline llvm/test/Transforms/PhaseOrdering/min_max_loop.ll:67]:
```suggestion
define i16 @vecreduce_smax_v2i16(i32 %n, ptr %v) {
```

**Review by dtcxzyw — COMMENTED** [review]
LGTM

File: llvm/test/Transforms/PhaseOrdering/min_max_loop.ll (hunk 9 above)
Comment (dtcxzyw) [reply llvm/test/Transforms/PhaseOrdering/min_max_loop.ll]: My command: `bin/opt -S -O3 -print-changed min_max.ll`
Then I got IR after the SROA pass :)

File: llvm/test/Transforms/PhaseOrdering/min_max_loop.ll (hunk 11 above)
Comment (leo-ard) [reply llvm/test/Transforms/PhaseOrdering/min_max_loop.ll]: oups, I missed those. Just fixed it:)

**Review by nikic — APPROVED** [review]
LGTM

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thank you!

Do you have the access to merge PR?

**Comment by leo-ard** [comment]
> Do you have the access to merge PR?

No I don't. 

Thanks for the time and insightful comments !
